components/org.wso2.carbon.identity.api.user.recovery/org.wso2.carbon.identity.rest.api.user.recovery.v2/src/main/java/org/wso2/carbon/identity/rest/api/user/recovery/v2/impl/core/PasswordRecoveryService.java

```diff
@@ -0,0 +1,463 @@
+import org.wso2.carbon.identity.governance.service.notification.NotificationChannels;
+import org.wso2.carbon.identity.recovery.IdentityRecoveryClientException;
+import org.wso2.carbon.identity.recovery.IdentityRecoveryConstants;
+import org.wso2.carbon.identity.recovery.IdentityRecoveryException;
+import org.wso2.carbon.identity.recovery.dto.PasswordRecoverDTO;
+import org.wso2.carbon.identity.recovery.dto.PasswordResetCodeDTO;
+import org.wso2.carbon.identity.recovery.dto.RecoveryChannelInfoDTO;
+import org.wso2.carbon.identity.recovery.dto.RecoveryInformationDTO;
+import org.wso2.carbon.identity.recovery.dto.ResendConfirmationDTO;
+import org.wso2.carbon.identity.recovery.dto.SuccessfulPasswordResetDTO;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.impl.core.utils.RecoveryUtil;
+
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.APICall;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.AccountRecoveryType;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.ConfirmRequest;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.InitRequest;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.PasswordRecoveryExternalNotifyResponse;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.PasswordRecoveryInternalNotifyResponse;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.PasswordResetResponse;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.RecoveryChannel;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.RecoveryChannelInformation;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.RecoveryRequest;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.ResendConfirmationCodeExternalResponse;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.ResendConfirmationCodeInternalResponse;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.ResendConfirmationRequest;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.ResetCodeResponse;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.ResetRequest;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.ws.rs.core.Response;
+
+/**
+ * Call internal OSGI services to perform password recovery.
+ */
+public class PasswordRecoveryService {
+
+    private static final Log log = LogFactory.getLog(PasswordRecoveryService.class.getName());
+
+    /**
+     * Initiate Password Recovery from POST.
+     *
+     * @param initRequest {@link InitRequest} object which holds the information in the account recovery
+     *                    request
+     * @return Response
+     */
+    public Response initiatePasswordRecovery(InitRequest initRequest) {
+
+        String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        Map<String, String> userClaims = RecoveryUtil.buildUserClaimsMap(initRequest.getClaims());
```

Comment:
what will happen if `initRequest` becomes null?

Reply:
Updated to throw a 400 error with 9817f1c6420aa1968231957e9f3114ec2253b03b